onmt/translate/BeamSearcher.lua

```diff
@@ -0,0 +1,395 @@
+
+--]]
+local function localize(v, x)
+  if x:type() == 'torch.CudaTensor' then
+    return v:cuda()
+  elseif x:type() == 'torch.CudaByteTensor' then
+    return v:cuda()
+  elseif x:type() == 'torch.CudaLongTensor' then
+    return v:cuda()
+  else
+    return v
+  end
+end
+
+--[[Constructor
+
+Parameters:
+
+  * `advancer` - an `onmt.translate.BeamSearchAdvancer` object.
+  * `endSymbol` - end symbol in the vocabulary. [onmt.Constants.EOS]
+
+]]
+function BeamSearcher:__init(advancer, endSymbol)
+  self.stepFunction = advancer.step
+  self.keptStateIndexes = advancer.keptStateIndexes or {}
+  self.endSymbol = endSymbol or onmt.Constants.EOS
+end
+
+--[[ Perform beam search.
+
+Parameters:
+
+  * `beamSize` - beam size. [1]
+  * `nBest` - the `nBest` top hypotheses can be returned after performing beam search. This value must be smaller than or equal to `beamSize`. [1]
+
+]]
+function BeamSearcher:search(beamSize, nBest)
+  assert (nBest <= beamSize)
+  self.nBest = nBest or 1
+  self.beamSize = beamSize or 1
+  self.origBatchIdToRemainingBatchId = {}
+  self.origBatchSize = nil
+  self.topIndexesHistory = {}
+  self.stepOutputsHistory = {}
+  self.beamParentsHistory = {}
+  self.beamScoresHistory = {}
+  self.completedTimeStep = {}
+
+  local vocabSize
+  local topIndexes
+  local beamScores
+  local stepOutputs, scores
+  local remainingBatchIdToOrigBatchId = {}
+
+  local t = 1
+  while true do
+    local flatTopIndexes = topIndexes
+    if flatTopIndexes then
+      flatTopIndexes = flatTopIndexes:view(-1)
+    end
+    -- Go one step forward
+    scores, stepOutputs = self.stepFunction(flatTopIndexes, stepOutputs)
+    if scores == nil then
+      self.maxSeqLength = t - 1
+      break
+    end
+    if vocabSize then
+      assert (vocabSize == scores:size(2))
+    else
+      vocabSize = scores:size(2)
+    end
+    -- Figure out the top k indexes, and which beam do they come from
+    local rawIndexes, remainingBatchSize
+    if t == 1 then
+      self.origBatchSize = scores:size(1)
+      remainingBatchSize = self.origBatchSize
+      for b = 1, self.origBatchSize do
+        remainingBatchIdToOrigBatchId[b] = b
+      end
+      beamScores, rawIndexes = scores:topk(self.beamSize, 2, true, true)
+      rawIndexes:add(-1)
+      topIndexes = localize(rawIndexes:double(), rawIndexes) + 1
+    else
+      remainingBatchSize = math.floor(scores:size(1) / self.beamSize)
+      -- Set other tokens scores to -inf to avoid ABCD<EOS>FG on beam
+      if self.nBest > 1 then
+        local minScore = -9e9
+        scores:add(localize(topIndexes:view(-1):eq(self.endSymbol):double(), scores)
+          :mul(minScore):view(-1, 1):expand(scores:size(1), vocabSize))
+      end
+      -- Ensure that tokens after <EOS> remain <EOS> and scores do not change
+      scores:select(2, self.endSymbol)
+        :maskedFill(topIndexes:view(-1):eq(self.endSymbol), 0)
+      local totalScores = (scores:view(remainingBatchSize, self.beamSize
```

Comment:
haha my vim does not come with a smart indentation so I did it manually. How many spaces would you suggest?